### ISSUE TITLE
Bad text rendering?

### ISSUE THREAD

**Issue by 0x14307** [slint-ui/slint#6298]
Hi there! 

I was working on a small personal project of mine (Rust, Windows x86_64) and noticed that text rendering is sort of botched where it just doesn't render properly and looks to have some artifacts with the actual executable itself whilst the Slint preview works just fine.

![image](https://github.com/user-attachments/assets/cd6ccb84-cf61-4abc-a802-adf6a40696d2)

This is using the default font that is provided to Slint by the system (so no font-families here) and will show up no-matter the application style (Fluent, Cupertino, etc.)

**Comment by tronical** [comment]
I think this is a consequence of #5177 :-(, as it emphasises the quality of the glyph rasterisation even more. Best path forward would probably be to move to cosmic-text, if even just for femtovg to start with.

**Comment by qarmin** [comment]
I also found that when using VerticalBox, but is quite strange the problem happens only when scrollbar is at the end/start, otherwise text is quite good

https://github.com/user-attachments/assets/fb794722-758d-4b7a-a646-eb658868fbc9

![Screenshot from 2024-10-06 12-16-23](https://github.com/user-attachments/assets/e5501301-0f33-4115-a219-385f60a8b9f2)
![Screenshot from 2024-10-06 12-16-37](https://github.com/user-attachments/assets/82833355-9547-4476-8d09-2da1ec1e5ae8)

**Comment by expenses** [comment]
Possibly related to #9310?

**Comment by LeonMatthes** [comment]
#10931 contains more discussion around this, as it seems to be a duplicate of this issue.

**Comment by LeonMatthes** [comment]
@tronical I think I know why this is happening 🙈 
If Text ends up at exactly a half-pixel location (so `X.5px`), the "nearest neighbor" is basically a gamble between the left/right pixel and depends on floating-point stability (which aren't terribly stable).

I have confirmed this by putting a debug in the draw_glyph_run function of the femtovg renderer. Every time the canvas is transformed at a .5 pixel I get the issue.
Let me try to come up with a patch for it.

**Comment by tronical** [comment]
Yeah that explains at least why the source texture is fine. It would be good to do what we do with Skia, exactly. Been wanting that but my inability to reproduce didn't help :(

**Comment by LeonMatthes** [comment]
@tronical I tried to come up with a fix in #11335. I don't know the FemtoVG API well, so let me know if there is a better way to do it.

The text is very crisp now in any case 🥳

**Comment by tronical** [comment]
> @tronical I tried to come up with a fix in #11335. I don't know the FemtoVG API well, so let me know if there is a better way to do it.
> 
> The text is very crisp now in any case 🥳 
> 
> 

That's an excellent fix. ❤️One detail missing. I'll comment on the PR.
